### PR TITLE
Expanding existing kuttl tests to cover case with multiple nodesets

### PR DESCRIPTION
**Pull request by jpodivin**
Tests addition of new nodeset into existing deployment and redeploying.

### CODE REVIEW

**Review by slagle — COMMENTED** [review]
Adding a NodeSet to a Deployment that is already deployed (Deployment.Status.Deployed=true) isn't going to do anything. You'd need to create a new Deployment instead.

I don't think your assert actually asserts that any OpenStackAnsibleEE's were created for the edpm-compute-beta-nodeset. So, nothing was actually deployed.

Looks like this did find a bug though, in that we shouldn't add the deployedConfigHash, ConfigMapHashes, and SecretHashes to the NodeSet if it's not actually deployed.

I think we need to update the Deployment reconciler to add NodeSetHashes to Deployment.Status. Then from the deploymentWatcher for the NodeSet, we can actually confirm if the Deployment deployed the actual NodeSet.Status.ConfigHash, before we copy any deployed hashes from the Deloyment to the NodeSet.

**Comment by jpodivin** [comment]
> Adding a NodeSet to a Deployment that is already deployed (Deployment.Status.Deployed=true) isn't going to do anything. You'd need to create a new Deployment instead.
> 
> I don't think your assert actually asserts that any OpenStackAnsibleEE's were created for the edpm-compute-beta-nodeset. So, nothing was actually deployed.
> 
> Looks like this did find a bug though, in that we shouldn't add the deployedConfigHash, ConfigMapHashes, and SecretHashes to the NodeSet if it's not actually deployed.
> 
> I think we need to update the Deployment reconciler to add NodeSetHashes to Deployment.Status. Then from the deploymentWatcher for the NodeSet, we can actually confirm if the Deployment deployed the actual NodeSet.Status.ConfigHash, before we copy any deployed hashes from the Deloyment to the NodeSet.

Right. The nodeset reaches all the conditions but the deployment is already done.

**Review by slagle — CHANGES_REQUESTED** [review]
The ansible execution for edpm-compute-beta-nodeset is not actually "done" though. So, the test you've added here tests a buggy behavior (and passes). That's not what we want.

**Comment by slagle** [comment]
> The ansible execution for edpm-compute-beta-nodeset is not actually "done" though. So, the test you've added here tests a buggy behavior (and passes). That's not what we want.

This would fix that behavior:
https://github.com/openstack-k8s-operators/dataplane-operator/pull/763
Can you give that a review and check my logic there? Let me know what you think.

**Comment by jpodivin** [comment]
> > The ansible execution for edpm-compute-beta-nodeset is not actually "done" though. So, the test you've added here tests a buggy behavior (and passes). That's not what we want.
> 
> This would fix that behavior: #763 Can you give that a review and check my logic there? Let me know what you think.

I've rewritten the test so that it now uses different deployment. So it should no longer get "deployed" due to bug.

However I think the better option would be to prevent addition of new nodesets into deployment outright, like @SeanMooney suggests.

**Review by slagle — APPROVED** [review]

**Comment by jpodivin** [comment]
/retest

Tests still appear to be flaky.

File: tests/kuttl/tests/dataplane-deploy-no-nodes-test/06-assert.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  - download-cache
+  - bootstrap
+  env:
+  - name: ANSIBLE_FORCE_COLOR
+    value: "True"
+  nodes: {}
+  nodeTemplate:
+    ansibleSSHPrivateKeySecret: dataplane-ansible-ssh-private-key-secret
+status:
+  conditions:
```
Comment (rabi) [inline tests/kuttl/tests/dataplane-deploy-no-nodes-test/06-assert.yaml:17]: You need to add `ServiceAccountReady` condition here too.

        +  - lastTransitionTime: "2024-03-22T08:26:47Z"
        +    message: ServiceAccount created
        +    reason: Ready
        +    status: "True"
        +    type: ServiceAccountReady

Comment (jpodivin) [reply tests/kuttl/tests/dataplane-deploy-no-nodes-test/06-assert.yaml:17]: Done, config hashes too.

**Comment by jpodivin** [comment]
recheck

```
timed out waiting for the condition on openstackdataplanenodesets/openstack-edpm-ipam
```

File: tests/kuttl/tests/dataplane-deploy-no-nodes-test/06-assert.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    type: InputReady
+  - message: ServiceAccount created
+    reason: Ready
+    status: "True"
+    type: ServiceAccountReady
+  - message: Setup complete
+    reason: Ready
+    status: "True"
+    type: SetupReady
+  configHash: n696hcdh665h56ch74hc4h559hc5hdch5cdh5b8hb5h694hd8h65bh696h56chfhc4hcbh64fh5b8h5b5h9dh58h8bh98h97hf6h8h5cdhf9q
```
Comment (rabi) [inline tests/kuttl/tests/dataplane-deploy-no-nodes-test/06-assert.yaml:38]: Is there any point of asserting these config hashes? I had removed them from other tests earlier as someone adding anything has to run the tests to find these hashes or calculate them manually.

**Review by rabi — COMMENTED** [review]
Can we also assert the AnsibleEE resources for the nodesets as we do in other tests?

File: tests/kuttl/tests/dataplane-deploy-no-nodes-test/06-assert.yaml (hunk 2 above)
Comment (jpodivin) [reply tests/kuttl/tests/dataplane-deploy-no-nodes-test/06-assert.yaml:38]: The point is exactly that. If the hashes change it means something has changed in our code. If the change is expected there is nothing easier than filling the hash. If not, it indicates a potential issue.

This way the developer can't ignore necessity of revising existing tests.

Comment (rabi) [reply tests/kuttl/tests/dataplane-deploy-no-nodes-test/06-assert.yaml:38]: Hashes are not meant for code change, but the spec changes after deployment.  It's only meant to compared when specs are updated.

Comment (jpodivin) [reply tests/kuttl/tests/dataplane-deploy-no-nodes-test/06-assert.yaml:38]: I know hashes are computed from the data. You don't have to tell me that. But we are asserting on the result.
If, hypothetically, something changes in the way operator works with spec, or there is a change in kubebuilder defaults or one of many other things, the hash will change. Indicating that something is happening.

And if that happens in PR that shouldn't change the hash, we will know that it's touching something it shouldn't touch.

**Comment by jpodivin** [comment]
Affected by ongoing CIX, do not recheck.

**Comment by jpodivin** [comment]
recheck - certmanager

**Review by slagle — APPROVED** [review]

**Pull request merged**